lisa/tests/scheduler/load_tracking.py

```diff
@@ -971,13 +977,19 @@ def get_trace_cpu_util(self):
             # Trim the end a bit, otherwise we could have one or two events
             # from the next phase
             end = phase_start + self.phases_durations[phase] * .9
-
             phase_df = df[start:end]
-            phase_duration = end - start
 
             for cpu in self.cpus:
                 util = phase_df[phase_df.cpu == cpu].util
-                cpu_util[cpu][phase] = series_integrate(util) / (phase_duration)
+                # The runqueue util signal's average does not match the duty
+                # cycle of the task, since it "decays instantly" at next task
+                # wakeup, but stays at its previous value when the task sleeps.
+                # This means that rq PELT signal average is higher than the
+                # idealized PELT signal. Using trapz integration allows to
+                # lower the contribution of the sleep-time util, since it links
+                # with a straight line the point when task goes to sleep with
+                # the wakeup util point.
+                cpu_util[cpu][phase] = series_mean(util, method='trapz')
```

Comment:
FWIW,  that's what BART was doing

https://pythonhosted.org/bart-py/bart.common.Utils.html

And I still noticed the approximation errors but I think for the purposes of all of our tests it's good enough. I did document them in an old PR with that noisy tasks if anyone cares to look at it in the future

Reply:
Yes, that was the default behavior of bart. In this case, we don't face an approximation error, as stated in the comment introduced by this PR. We use trapz to workaround the fact that the real mean of the signal we look at is not the mean of the utilization of the task, regardless of how many sample points we would take on that signal. It's a bit brittle and the error is not really controlled, but there is no better/more robust alternative ATM ...

FYI `lisa.datautils.series_integrate()` uses the rectangle method by default since that is the correct thing to do to integrate almost all signals we manipulate, with the exception of the power probe measurements. All trace signals we have are 100% accurate, since we get an event as soon as their value is updated (i.e. watching the value of the signal every picosecond would not expose more information than we already have from the trace).

Reply:
Hmm okay. I didn't dig into the details with the newer code. But I think the error that I encountered in bart was due to the linear approximation of the decay - which is exponential in reality. The scheduler doesn't record any samples during the decay. So we end up with a slightly higher util_mean than the real value. The significance of this error will depend on the duty cycle and the ratio of run/sleep. We didn't do much about at the end because it didn't turn out to be a real problem and the noisy tasks were the real culprit. Then buildroot solved the noise issue - mostly at least.

Anyway, I have nothing against the new change - just trying to add some references/docs for the future.

Reply:
Yes, that's exactly what is now described in the comment, and the reason why I forced the task period to a known value in that class (changing it would break that hack)

Reply:
Okay. I didn't parse the comment that way tbh. But as long as we are talking about the same thing it's all good :)